network/p2p/dns/resolver.go

```diff
@@ -74,17 +93,37 @@ func (r *Resolver) LookupIPAddr(ctx context.Context, domain string) ([]net.IPAdd
 }
 
 // lookupIPAddr encapsulates the logic of resolving an ip address through cache.
+// If domain exists on cache it is resolved through the cache.
+// An expired domain on cache is still addressed through the cache, however, a request is fired up asynchronously
+// through the underlying basic resolver to resolve it from the network.
 func (r *Resolver) lookupIPAddr(ctx context.Context, domain string) ([]net.IPAddr, error) {
-	if addr, exits, invalidated := r.c.resolveIPCache(domain); exits {
-		// resolving address from cache
-		r.collector.OnDNSCacheHit()
-		return addr, nil
-	} else if invalidated {
-		r.collector.OnDNSCacheInvalidated()
+	addr, exists, fresh := r.c.resolveIPCache(domain)
+
+	if !exists {
+		r.collector.OnDNSCacheMiss()
+		return r.lookupResolverForIPAddr(ctx, domain)
 	}
 
-	// resolves domain through underlying resolver
-	r.collector.OnDNSCacheMiss()
+	if !fresh && r.shouldResolveIP(domain) {
+		r.unit.Launch(func() {
+			_, err := r.lookupResolverForIPAddr(ctx, domain)
+			if err != nil {
+				// invalidates cached entry when hits error on resolving.
+				invalidated := r.c.invalidateIPCacheEntry(domain)
```

Comment:
> Why invalidate the cache upon error when resolving? Wouldn't we just flag look-up again? Or forget about it because we made a best effort and did one lookup. Any DNS lookup could easily fail due to a missing UDP packet? But does that mean we should invalidate the cache?

Thanks for the comment, I appreciate the optimization suggestion, but I think we need to be careful about the safety margins we move. As mentioned to an [earlier comment](https://github.com/onflow/flow-go/pull/1216#discussion_r714242218), we'd safer to reflect error directly to application layer and delete the expired entry as soon as we hit an error. The cache is in place to boost up the _happy path_ execution, and currently is not designed for handling unhappy path scenarios. Hence, at the current version, a fail-safe cache that boosts up the happy path suites more secure than a cache trying to deal with DNS errors. Of course, you comment is valuable to consider and I will issue it up to craft a well-thought solution later. 

> I think there are two types of lookup errors here:
> 
> 1. Lookup failed due to a trivial reason such as UDP packet going missing.
> 2. Lookup failed due to a major reason such as infrastructure re-configured during epoch.

Your thought are pretty valid on a theoretical perspective. However, digging a bit into the underlying resolver, I found that it resides on top of several dependencies ranging from libp2p to native Go, and there is no such strict guarantee on the returned error types through its dependency injection. In short, an error means unhappy path without any guarantee on the type. So, unless we implement a wrapper around the underlying DNS to cast the errors into well-defined error types that we can distinguish case (1) and (2) from any other default case, it does not look safe to establish the assumption on implementation. I issued up https://github.com/dapperlabs/flow-go/issues/5872 to re-surface your suggestion later on.

Reply:
Thanks for the comments!

Yes, I'm not convinced that the unhappy path has been thought through, or that this PR is necessarily taking the best action upon hitting the unhappy path by invalidating the cache entry. For a single `NXDOMAIN` then this will result in not one but two errors (the last cached IP result is returned which presumably does not work anymore, and later the domain will be invalidated in the cache which presumably generates a 2nd error) which might be confusing for the person debugging.

However, I'm happy to approve this PR as is, assuming that the unhappy path is debated and possibly refined later using the issue [1] kindly created :-)

[1] https://github.com/dapperlabs/flow-go/issues/5872